dp/jobqueueing/gcsub.py

```diff
@@ -0,0 +1,164 @@
+"""
+Script to dequeue one or more generate_climos queue entries with NEW status, and submit a PBS job for each,
+updating the queue entries accordingly.
+
+Entries are dequeued in order of addition to the database; i.e., it is a FIFO queue.
+"""
+
+from argparse import ArgumentParser
+import datetime
+import logging
+import os.path
+import sys
+from subprocess import Popen, PIPE
+
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy import asc
+
+from dp.script_helpers import default_logger
+from dp.jobqueueing.argparse_helpers import add_global_arguments, add_submit_arguments
+from dp.jobqueueing.jobqueueing_db import GenerateClimosQueueEntry
+
+
+logger = default_logger()
+
+
+def make_qsub_test_script(queue_entry):
+    return '''
+#PBS -l nodes=1:ppn={qe.ppn}
+#PBS -l vmem={vmem}mb
+#PBS -l walltime={qe.walltime}
+#PBS -o {qe.output_directory}/logs
+#PBS -e {qe.output_directory}/logs
+#PBS -m abe
+#PBS -N generate_climos:{input_filename}
+#PBS -d /storage/home/rglover/code/climate-explorer-backend
+
+pbs_job_num=$(expr match "$PBS_JOBID" '\([0-9]*\)')
+
+# Set up the execution environment
+module load netcdf-bin
+module load cdo-bin
+source py3.4/bin/activate
```

Comment:
My thought was that we'd establish a standard environment independently of the scripts that use it. But you are right that it is fragile.

The naive approach would be to set up a new environment for every run of the script. It could even include a git pull on the repo to ensure that we are running the latest greatest. But it has the following issues to be addressed:

- Concurrency: These jobs only require one processor each, so typically 4 run at a time. If each script is running setup, they will interfere with each other. They'll have to establish their own uniquely named virtual env to run in. Alternatively we clone the whole repo to a uniquely named directory.
- Speed: There will be overhead for setting up each time. Not sure how much.
- Trickiness, time to dev: I recall that setting up the environment took some hacking about each time. Possibly due to my own shortcomings ... but a clean, error-free setup script has to be established and that will also take some time.